### PR TITLE
Add alpine support

### PR DESCRIPTION
**Pull request by at-wat**
This PR adds support for Alpine Linux apk package manager.

~It's currently not ready to be merged, since it requires latest (not yet released) rospkg with alpine detection.~ The change on rospkg has been merged and released.

I have tested `rosdep install` with forked rosdistro with some alpine package definitions, in docker alpine 3.7. (https://github.com/at-wat/alpine-ros)

### CODE REVIEW

File: src/rosdep2/platforms/alpine.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python
+# Copyright (c) 2018, rosdep authors
```
Comment (tfoote) [inline src/rosdep2/platforms/alpine.py]: 'rosdep authors' is not a valid copyright entity. If you want to assign the copyright please use the foundation, Open Source Robotics Foundation

File: test/test_rosdep_alpine.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# Copyright (c) 2018, rosdep authors
```
Comment (tfoote) [inline test/test_rosdep_alpine.py]: same here

**Comment by nuclearsandwich** [comment]
I have a couple of alpine systems, one physical, one an lxc container, that I can review this on.

**Comment by tfoote** [comment]
@Minipada if you could test this it would be great too. Thanks

File: src/rosdep2/platforms/alpine.py (hunk 1 above)
Comment (at-wat) [reply src/rosdep2/platforms/alpine.py]: I changed it to:
```
# Copyright (c) 2018, Open Source Robotics Foundation, Inc.
# Copyright (c) 2018, SEQSENSE, Inc.
```
Is it OK?

Comment (at-wat) [reply src/rosdep2/platforms/alpine.py]: OK, thanks for your suggestion.

**Comment by Minipada** [comment]
I just tested your container, everything is working well.
**How I tested**:

Check one modified key is resolved (your patch [here](https://github.com/at-wat/rosdistro/commit/069e74f372277b11fafc5f00653cb8cdb6390ddb))
```
/test_ws # rosdep resolve apr
#apk
apr apr-util
```

Check one can install from rosdep:

```
/test_ws # cat src/package.xml 
<?xml version="1.0"?>
<package format="2">
  <name>sample_pkg</name>
  <version>0.0.0</version>
  <description>sample package</description>

  <maintainer email="atsushi.w@ieee.org">Atsushi Watanabe</maintainer>
  <license>BSD</license>
  <author email="atsushi.w@ieee.org">Atsushi Watanabe</author>

  <depend>tinyxml</depend>
</package>



/test_ws # apk del tinyxml-dev
(1/4) Purging tinyxml-dev (2.6.2-r1)
(2/4) Purging tinyxml (2.6.2-r1)
(3/4) Purging libstdc++ (6.4.0-r5)
(4/4) Purging libgcc (6.4.0-r5)
OK: 62 MiB in 25 packages

/test_ws # rosdep install -y --from-paths src
executing command [apk add tinyxml-dev]
(1/4) Installing libgcc (6.4.0-r5)
(2/4) Installing libstdc++ (6.4.0-r5)
(3/4) Installing tinyxml (2.6.2-r1)
(4/4) Installing tinyxml-dev (2.6.2-r1)
OK: 63 MiB in 29 packages
#All required rosdeps installed successfully
```

**Comment by at-wat** [comment]
Latest rospkg with Alpine Linux detection is released.
This PR should now pass CI test.
@tfoote Could you re-trigger Travis build?

**Comment by tfoote** [comment]
@at-wat It looks to be failing on OSX due to a test config error something must have moved in the osx packaging.
```
 nosetests --with-coverage --cover-package=rosdep2 --with-xunit test
/Users/travis/.travis/job_stages: line 78: nosetests: command not found
```

**Comment by at-wat** [comment]
@tfoote test config problem seems fixed by #629. Now, this should pass all tests.

**Comment by nuclearsandwich** [comment]
@at-wat in order for your PR to benefit from the changes in #629 it needs to be updated to include those features. I've opened a PR on your PR https://github.com/at-wat/rosdep/pull/1 to merge in the latest commits from master. Merging that if it is to your liking should re-trigger CI.

**Comment by at-wat** [comment]
Oh, I've forgotten to merge master into this. Thank you @nuclearsandwich

File: src/rosdep2/platforms/alpine.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python
```
Comment (at-wat) [inline src/rosdep2/platforms/alpine.py]: shebang is not needed here (https://github.com/ros-infrastructure/rosdep/pull/630)

**Comment by russkel** [comment]
What's holding this up being merged? Let me know how I can help.

**Comment by russkel** [comment]
There seems to be some conflicts between the pip installed PyYAML and the apk installed PyYAML. It causes the rosdep installing of dependencies to fail.

If py2-yaml is installed before it is installed by apk then we get this failure.

```
ERROR: the following rosdeps failed to install
  apk: command [apk add apr apr-util tinyxml2-dev lz4-dev py-nose py-imaging sbcl py-yaml bzip2-dev cmake] failed
```
```
(28/31) Installing py2-yaml (5.1-r0)
ERROR: py2-yaml-5.1-r0: failed to rename usr/lib/python2.7/site-packages/.apk.f759bf6f1398870b2e65d7d6bebb047f9911fdf8b72b5bed to usr/lib/python2.7/site-packages/PyYAML-5.1-py2.7.egg-info.
```

Also, should we just opt for python3 by default in adding the alpine platform?

**Comment by russkel** [comment]
In addition to this, the newest version of boost has removed the deprecated boost-signals library, so this fails on building roscpp.

**Comment by at-wat** [comment]
@russkel the problems around PyYAML and boost-signals are independent of this PR.
Mixing pip and distribution native package manager can cause conflict not only on Alpine Linux. Especially, Alpine doesn't have a lot of packages on native package repository because of their policy to keep it light-weighted. I'm maintaining third-party Alpine package repository to provide such kind of packages (https://github.com/alpine-ros/alpine-ros, https://github.com/seqsense/aports-ros-experimental). Please try https://github.com/alpine-ros/alpine-ros docker image which uses our public repository providing many ported/backported packages compatible with kinetic.
Valid pair of boost version and ROS distro is limited. (It's also not Alpine specific problem.) I confirmed that boost and gcc on Alpine 3.7 are compatible with kinetic and 3.8 looks compatible with melodic.

@tfoote @nuclearsandwich Our company is already using ROS on Alpine Linux. `rosdep` in this PR looks working correctly. If you need any additional information to go forward, please let me know.

**Comment by russkel** [comment]
> @russkel the problems around PyYAML and boost-signals are independent of this PR.
Yep I understand this. I just wanted to document the issues one would have trying to bring up a ROS melodic setup with this (and I am trying to work on them so I can get melodic running in docker).

Thanks for your repositories! Good to know you are having success with Alpine at your company.

**Pull request merged**